### PR TITLE
Core/ts version task 357 balance api

### PR DESCRIPTION
**Pull request by WilliamHawkingStephen**
## Check list  

- [x] All Merge Request pipelines should be passed.
- [x] Make sure all the changes are necessary and correct.
- [x] Make sure no needed change missing (all necessary changes are made).
- [x] Merge Request should contains only changes for one purpose (one feature or bug or refactoring).
- [x] All necessary unit tests are written.
- [x] Critical changes should be tested locally (or on staging).

### CODE REVIEW

File: ts-version/src/presentation/http/controllers/balance.controller.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  constructor(private dataService: IDataService.schema) {
+    super();
+  }
+
+  public getByUser = async (req: Request, res: Response): Promise<void> => {
+    const { user_address, token_address, network } = req.query;
+
+    const provider = Providers.getInstance().getProvider(
+      network as string,
+    ) as IBlockchain;
```
Comment (GreatGaryCooper) [inline ts-version/src/presentation/http/controllers/balance.controller.ts:18]: As far as I know, there is no need to use `as IBlockchain` since the return type of the `getProvider` method is already `IBlockchain`.

File: ts-version/src/presentation/http/express.ts [hunk 2]
```diff
@@ -132,6 +134,12 @@ export default class HttpServer {
       (req: Request, res: Response, next: NextFunction) =>
         this.orderController.cancelAll(req, res).catch(next),
     );
+    this.expressApp.get(
+      "/user/balance",
+      auth,
+      ...HTTPValidator.getBalanceByUser,
```
Comment (GreatGaryCooper) [inline ts-version/src/presentation/http/express.ts:140]: Based on the new changes(they are not released yet) we should first validate the parameters and then authenticate the user since there are some rules in the `auth` middleware that uses the provided http parameters.
So please swap `auth` and  `...HTTPValidator.getBalanceByUser` order.

File: ts-version/src/presentation/http/controllers/balance.controller.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    res.send({ success: true, allowance });
+  };
+
+  public getPossession = async (req: Request, res: Response): Promise<void> => {
+    const { user_address, network, market } = req.query;
+
+    const provider = Providers.getInstance().getProvider(
+      network as string,
+    ) as IBlockchain;
```
Comment (GreatGaryCooper) [inline ts-version/src/presentation/http/controllers/balance.controller.ts:58]: Same here, no need to convert to `IBlockchain`

**Review by GreatGaryCooper — APPROVED** [review]
Looks good to me.
Please also add a couple of endpoint tests if you had time, just mock the provider calls and that's all.
If you don't have time for it right now, create a card to implement it later.

Thanks a lot buddy 🍻

File: ts-version/src/presentation/http/controllers/balance.controller.ts (hunk 1 above)
Comment (WilliamHawkingStephen) [reply ts-version/src/presentation/http/controllers/balance.controller.ts:18]: thanks a lot broski :hand:  
because of return value of getProvider which is : 
`IBlockchain | undefined`

but thank you about suggestion ill change it to check for undefined for getProvider

File: ts-version/src/utils/networks.ts [hunk 4]
```diff
@@ -11,7 +12,9 @@ export class Networks {
     const networks = await networkRepo.getNetworks();
     for (const network of networks) {
       const provider = Providers.getInstance().getProvider(network.name);
-      if (provider) await Network.addNetwork(network, provider, marketRepo);
+      if (!provider)
+        throw new ThirdPartyError(new Error(`provider(${network}) not found`));
```
Comment (GreatGaryCooper) [inline ts-version/src/utils/networks.ts:16]: Since `ThirdPartError` has it's isOperational value set to true, and the init function will be called when the app is starting and not by user requests, we should only throw an error, so that app will crash and we get to know the crash reason.
But if we throw the ThirdPartyError,since the isOperational property is set to true, we just log the error and continue the app flow, which is not what we want.

In simple terms, when built in error objects reach to our error handler, the system won't continue and process exits.

File: ts-version/src/test/e2e/networks.e2e.test.ts [hunk 5]
```diff
@@ -11,6 +11,7 @@ describe("GET /networks endpoint", () => {
   beforeAll(async () => {
     spiedSetInterval.mockImplementation(jest.fn());
     providers = Providers.getInstance();
+    console.log("provider", providers);
```
Comment (GreatGaryCooper) [inline ts-version/src/test/e2e/networks.e2e.test.ts:14]: Please remove this `console.log` statement.

**Review by GreatGaryCooper — APPROVED** [review]
LGTM

File: ts-version/src/presentation/http/controllers/balance.controller.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    super();
+  }
+
+  public getByUser = async (req: Request, res: Response): Promise<void> => {
+    const { user_address, token_address, network } = req.query;
+
+    const provider = Providers.getInstance().getProvider(network as string);
+    if (!provider)
+      throw new ThirdPartyError(new Error(`provider(${network}) not found`));
+    const etherBalance = await provider.getBalance(user_address as string);
```
Comment (hypersolo) [inline ts-version/src/presentation/http/controllers/balance.controller.ts:19]: use nativeCoinBalance instead of ether.
